### PR TITLE
Fix column __init__ args typing for EventSchemaTarget

### PR DESCRIPTION
**Pull request by MaicoTimmerman**
### Description
Fix `Column.__init__` for `*args`
### Checklist
Test-case:
```py
from sqlalchemy import Column, ForeignKey, Sequence
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import registry, Mapped

reg: registry = registry()


@reg.mapped
class InferredNames:
    __tablename__ = "infered"

    # Instantiated type
    id = Column(Integer(), primary_key=True)

    # Non-instantiated type
    name = Column(String, nullable=False)

    # SchemaItem
    foreign: Mapped[int] = Column(ForeignKey("infered.id"))

    # SchemaItem without type
    foreign1: Mapped[int] = Column(ForeignKey("infered.id"))

    # SchemaItem without type
    foreign2: Mapped[int] = Column(Integer, ForeignKey("infered.id"))

    # sequence
    seq = Column(Integer, Sequence("my_sequence"), primary_key=True)

    # type is None
    # TODO This still crashes the plugin, even though the docstring says is valid
    # Column(None, ForeignKey("addresses.id")),


@reg.mapped
class ExplicitNames:
    __tablename__ = "explicit"

    # Instantiated type
    id = Column("id", Integer(), primary_key=True)

    # Non-instantiated type
    name = Column("name", String, nullable=False)

    # SchemaItem without type
    foreign1: int = Column("foreign", ForeignKey("infered.id"))

    # SchemaItem without type
    foreign2: int = Column("foreign", Integer, ForeignKey("infered.id"))

    # sequence
    seq = Column(Integer, Sequence("my_sequence"), primary_key=True)

    # type is None
    # TODO This still crashes the plugin, even though the docstring says is valid
    # none_type = Column("none_type", None, ForeignKey("addresses.id"))
```

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by MaicoTimmerman** [comment]
Validated with `python -m mypy.stubtest sqlalchemy.sql.schema` and the test-cases in `/sqlalchemy/test/ext/mypy/files`

File: sqlalchemy-stubs/sql/schema.pyi
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/schema.pyi:139]: There are a couple of overloads that are missing. In total, we need to handle the following:

* `self, name: str, type_: Union[_TE, Type[_TE]], *args: SchemaEventTarget, <kwargs you added>`
* `self: Column[sqltypes.NullType], name: str, type_: None, *args: SchemaEventTarget, <kwargs you added>`
* `self, type_: Union[_TE, Type[_TE]], *args: SchemaEventTarget, name: str = ..., <kwargs you added>`
* `self: Column[sqltypes.NullType], type_: None, *args: SchemaEventTarget, name: str = ..., <kwargs you added>`
* `self, *args: SchemaEventTarget, name: str = ..., type_: Union[_TE, Type[_TE]], <kwargs you added>`
* `self: Column[sqltypes.NullType], *args: SchemaEventTarget, name: str = ..., type_: None = ..., <kwargs you added>`

File: sqlalchemy-stubs/sql/schema.pyi
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/schema.pyi:8]: This needs to be imported from `typing_extensions`

File: sqlalchemy-stubs/sql/schema.pyi
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/schema.pyi:139]: I missed two overloads (which I would put after the second overload above:

* `self, name: str, *args: SchemaEventTarget, type_: Union[_TE, Type[_TE]], <kwargs you added>`
* self: Column[sqltypes.NullType], name: str, *args: SchemaEventTarget, type_: None = ..., <kwargs you added>`

Comment (bryanforbes) [reply sqlalchemy-stubs/sql/schema.pyi:139]: The following should all pass type checking as well as produce the correct `Column` type:

```python
reveal_type(Column('asdf', Integer, ForeignKey('asdf'), index=True))  # Column[Integer]
reveal_type(Column('asdf', None, ForeignKey('asdf'), index=True))  # Column[NullType]
reveal_type(Column('asdf', ForeignKey('asdf'), index=True))  # Column[NullType]
reveal_type(Column('asdf', ForeignKey('asdf'), type_=Integer, index=True))  # Column[Integer]
reveal_type(Column(Integer, ForeignKey('asdf'), index=True))  # Column[Integer]
reveal_type(Column(None, ForeignKey('asdf'), index=True))  # Column[NullType]
reveal_type(Column(ForeignKey('asdf'), type_=Integer, name='asdf', index=True))  # Column[Integer]
reveal_type(Column(ForeignKey('asdf'), name='asdf', index=True))  # Column[NullType]
reveal_type(Column(ForeignKey('asdf'), index=True))  # Column[NullType]
reveal_type(Column(ForeignKey('asdf'), type_=None, index=True))  # Column[NullType]
reveal_type(Column(Integer, ForeignKey('asdf'), type_=None, index=True))  # Column[Integer]
reveal_type(Column('asdf', Integer, ForeignKey('asdf'), name='foo', type_=None, index=True))  # Column[Integer]
reveal_type(Column(None, ForeignKey('asdf'), type_=Integer, index=True))  # Column[NullType]
reveal_type(Column('asdf', None, ForeignKey('asdf'), name='foo', type_=Integer, index=True))  # Column[NullType]
```

Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/schema.pyi:139]: I programmatically generated all possible valid combinations and came to the following test-case:

```py
ffrom sqlalchemy import Column, ForeignKey, Integer, String

Column("name", Integer, index=True)
Column(None, name='name')
Column(Integer, name='name', index=True)
Column("name", ForeignKey("a.id"))
Column(ForeignKey("a.id"), type_=None, index=True)
Column(ForeignKey("a.id"), name='name', type_=Integer())
Column("name", None)
Column("name", index=True)
Column(ForeignKey("a.id"), name='name', index=True)
Column(type_=None, index=True)
Column(None, ForeignKey("a.id"))
Column("name")
Column(name='name', type_=None, index=True)
Column(ForeignKey("a.id"), name='name', type_=None)
Column(Integer)
Column(ForeignKey("a.id"), type_=Integer())
Column("name", Integer, ForeignKey("a.id"), index=True)
Column("name", None, ForeignKey("a.id"), index=True)
Column(ForeignKey("a.id"), index=True)
Column("name", Integer)
Column(Integer, name='name')
Column(Integer, ForeignKey("a.id"), name='name', index=True)
Column(ForeignKey("a.id"), type_=None)
Column(ForeignKey("a.id"), name='name')
Column(name='name', index=True)
Column(type_=None)
Column(None, index=True)
Column(name='name', type_=None)
Column(type_=Integer(), index=True)
Column("name", Integer, ForeignKey("a.id"))
Column(name='name', type_=Integer(), index=True)
Column(Integer, ForeignKey("a.id"), index=True)
Column("name", None, ForeignKey("a.id"))
Column(index=True)
Column("name", type_=None, index=True)
Column("name", ForeignKey("a.id"), type_=Integer(), index=True)
Column(ForeignKey("a.id"))
Column(Integer, ForeignKey("a.id"))
Column(Integer, ForeignKey("a.id"), name='name')
Column("name", ForeignKey("a.id"), index=True)
Column("name", type_=Integer(), index=True)
Column(ForeignKey("a.id"), name='name', type_=Integer(), index=True)
Column(name='name')
Column("name", None, index=True)
Column("name", ForeignKey("a.id"), type_=None, index=True)
Column("name", type_=Integer())
Column(None)
Column(None, ForeignKey("a.id"), index=True)
Column("name", ForeignKey("a.id"), type_=None)
Column(type_=Integer())
Column(None, ForeignKey("a.id"), name='name', index=True)
Column(Integer, index=True)
Column(ForeignKey("a.id"), name='name', type_=None, index=True)
Column(ForeignKey("a.id"), type_=Integer(), index=True)
Column(name='name', type_=Integer())
Column(None, name='name', index=True)
Column()
Column(None, ForeignKey("a.id"), name='name')
Column("name", type_=None)
Column("name", ForeignKey("a.id"), type_=Integer())

# EXPECTED_MYPY: No overload variant of "Column" matches argument types "Type[Integer]", "ForeignKey", "Type[String]"
Column(Integer, ForeignKey("a.id"), type_=String)

# EXPECTED_MYPY: No overload variant of "Column" matches argument types "str", "ForeignKey", "str"
Column("name", ForeignKey("a.id"), name="String")
```

All of these type-check and have valid column types.

Comment (CaselIT) [reply sqlalchemy-stubs/sql/schema.pyi:139]: Nice idea!

I think this could go in a test file (maybe also with the code that generates it in a comment or function)

Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/schema.pyi:139]: Thanks! I saw you working on the test environment for this repo. Shall I include a cleaned up version of the generating code with the output in this repo?. Or put it up for review in gerrit for `sqlalchemy/test/ext/mypy/files/...` after this has been merged?

Comment (CaselIT) [reply sqlalchemy-stubs/sql/schema.pyi:139]: I'm actually not sure.

On one end having test files in two places it's a bit of a mess, on the other it makes sense to add test cases or modify some while changing the stubs here. I guess we should decide what to do at some point.

It may make sense committing the file here in the meantime. At worst we can move it later.

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:139]: many of these are not usage patterns I want to support.   Essentially "name" and "type_" should never be passed as kw arguments.    Beyond that, I'm not totally comfortable with explicit None.   But right off, all the below patterns should not be allowed and if we can we should disallow them in the code:


> Column(None, name='name')
> Column(Integer, name='name', index=True)
> Column(ForeignKey("a.id"), type_=None, index=True)
> Column(ForeignKey("a.id"), name='name', type_=Integer())
> Column("name", None)
> Column(ForeignKey("a.id"), name='name', index=True)
> Column(type_=None, index=True)
> Column(name='name', type_=None, index=True)
> Column(ForeignKey("a.id"), name='name', type_=None)
> Column(ForeignKey("a.id"), type_=Integer())
> Column(ForeignKey("a.id"), index=True)
> Column(Integer, name='name')
> Column(Integer, ForeignKey("a.id"), name='name', index=True)
> Column(ForeignKey("a.id"), type_=None)
> Column(ForeignKey("a.id"), name='name')
> Column(name='name', index=True)
> Column(type_=None)
> Column(name='name', type_=None)
> Column(type_=Integer(), index=True)
> Column(name='name', type_=Integer(), index=True)
> Column("name", type_=None, index=True)
> Column("name", ForeignKey("a.id"), type_=Integer(), index=True)
> Column(Integer, ForeignKey("a.id"), name='name')
> Column("name", type_=Integer(), index=True)
> Column(ForeignKey("a.id"), name='name', type_=Integer(), index=True)
> Column(name='name')
> Column("name", ForeignKey("a.id"), type_=None, index=True)
> Column("name", type_=Integer())
> Column(None, ForeignKey("a.id"), index=True)
> Column("name", ForeignKey("a.id"), type_=None)
> Column(type_=Integer())
> Column(None, ForeignKey("a.id"), name='name', index=True)
> Column(ForeignKey("a.id"), name='name', type_=None, index=True)
> Column(ForeignKey("a.id"), type_=Integer(), index=True)
> Column(name='name', type_=Integer())
> Column(None, name='name', index=True)
> Column(None, ForeignKey("a.id"), name='name')
> Column("name", type_=None)
> Column("name", ForeignKey("a.id"), type_=Integer())

then a bunch of use cases that I find to be "iffy", but might need to be acceped from a pep484 perspective:

> Column(None, ForeignKey("a.id"))
> Column("name")
> Column("name", None, ForeignKey("a.id"), index=True)
> Column(None, index=True)
> Column("name", None, ForeignKey("a.id"))
> Column(index=True)
> Column("name", None, index=True)
> Column(None)
> Column(Integer, index=True)
> Column()

Above, from a SQLAlchemy perspective you can't have a Column without a type that does not also have a ForeignKey constraint, so a lot of those are instant error message (hopefully).   but this might be more of a runtime issue.

I also find the explicit None kind of distasteful but im not sure it should be disallowed.

Comment (MaicoTimmerman) [reply sqlalchemy-stubs/sql/schema.pyi:139]: The proposed types in the PR just reflect what currently happens in the runtime. I personally think it is beter to have types reflect the runtime behavior. So if we limit the types, I'd suggest adjusting runtime with some deprecations for 2.0.

> Above, from a SQLAlchemy perspective you can't have a Column without a type that does not also have a ForeignKey constraint, so a lot of those are instant error message (hopefully). but this might be more of a runtime issue.

I don't think this is easily possible as the `ForeignKey` can be any of the `*args`. So that would have to be a runtime check. I can add those checks in runtime. 

> I also find the explicit None kind of distasteful but im not sure it should be disallowed.

I also do not see the benefit of explicit positional `None`, however currently this is also present in the test-cases and therefor I think they might be better left in:
- sqlalchemy/test/orm/inheritance/test_basic.py 
- sqlalchemy/sqlalchemy/test/orm/_fixtures.py

Before I start working on it, can we agree on the actions to take for this PR:
- Limiting the typing to not allow `name` and `type_` 
- Leave `None` a valid positional parameter, as it is used in the tests as well.
- Introduce `RemovedIn20Warning` in the 1.4 releases for the `kwargs` at runtime.
- Add runtime check for missing positional argument at runtime without `ForeignKey`.

Comment (zzzeek) [reply sqlalchemy-stubs/sql/schema.pyi:139]: > The proposed types in the PR just reflect what currently happens in the runtime. I personally think it is beter to have types reflect the runtime behavior. So if we limit the types, I'd suggest adjusting runtime with some deprecations for 2.0.

that was my proposal, except these would be 1.x deprecations.    my concern is that the stubs are going to encourage, in IDEs, incorrect patterns.    


> 
> > Above, from a SQLAlchemy perspective you can't have a Column without a type that does not also have a ForeignKey constraint, so a lot of those are instant error message (hopefully). but this might be more of a runtime issue.
> 
> I don't think this is easily possible as the `ForeignKey` can be any of the `*args`. So that would have to be a runtime check. I can add those checks in runtime.

It might be interpreting None as a type with NullType which will eventually raise an error when someone tries to CREATE TABLE w/ the column and that might be where we have to leave it.


> 
> > I also find the explicit None kind of distasteful but im not sure it should be disallowed.
> 
> I also do not see the benefit of explicit positional `None`, however currently this is also present in the test-cases and therefor I think they might be better left in:
> 
>     * sqlalchemy/test/orm/inheritance/test_basic.py
> 
>     * sqlalchemy/sqlalchemy/test/orm/_fixtures.py

those fixtures are just wrong and those were probably written when the "can omit type if there's an FK" case was first created.    Positional "None" clearly has to stay functional for now but this goes under the same realm of things to deprecate, then the issue of the stubs encouraging patterns that I never intended people to use.

> 
> 
> Before I start working on it, can we agree on the actions to take for this PR:
> 
>     * Limiting the typing to not allow `name` and `type_`
> 
>     * Leave `None` a valid positional parameter, as it is used in the tests as well.
> 
>     * Introduce `RemovedIn20Warning` in the 1.4 releases for the `kwargs` at runtime.

regular SADeprecationWarning.   shouldnt be dependent on "testing for 2.0" mode.

> 
>     * Add runtime check for missing positional argument at runtime without `ForeignKey`.

this might be a valid case, assuming the None for type becomes NullType.

File: sqlalchemy-stubs/sql/elements.pyi
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/elements.pyi:428]: This is taken care of in #37

File: sqlalchemy-stubs/sql/schema.pyi
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/schema.pyi:209]: This only needs to be `self`

File: sqlalchemy-stubs/sql/schema.pyi
Comment (bryanforbes) [inline sqlalchemy-stubs/sql/schema.pyi:186]: This only needs to be `self`

**Review by CaselIT — APPROVED** [review]

**Pull request merged**
